llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp

```diff
@@ -0,0 +1,124 @@
+//===-- LowerWGLocalMemory.cpp - SYCL kernel local memory allocation pass -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass replaces calls to __sycl_allocateLocalMemory(Size, Alignment)
+// function with allocation of memory in local address space at the kernel
+// scope.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/LowerWGLocalMemory.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "LowerWGLocalMemory"
+
+static constexpr char SYCL_ALLOCLOCALMEM_CALL[] = "__sycl_allocateLocalMemory";
+static constexpr char LOCALMEMORY_GV_PREF[] = "WGLocalMem";
+
+namespace {
+class SYCLLowerWGLocalMemoryLegacy : public ModulePass {
+public:
+  static char ID;
+
+  SYCLLowerWGLocalMemoryLegacy() : ModulePass(ID) {
+    initializeSYCLLowerWGLocalMemoryLegacyPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager DummyMAM;
+    auto PA = Impl.run(M, DummyMAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLLowerWGLocalMemoryPass Impl;
+};
+} // namespace
+
+char SYCLLowerWGLocalMemoryLegacy::ID = 0;
+INITIALIZE_PASS(SYCLLowerWGLocalMemoryLegacy, "sycllowerwglocalmemory",
+                "Replace __sycl_allocateLocalMemory with allocation of memory "
+                "in local address space",
+                false, false)
+
+ModulePass *llvm::createSYCLLowerWGLocalMemoryPass() {
+  return new SYCLLowerWGLocalMemoryLegacy();
+}
+
+static bool lowerAllocaLocalMem(Module &M) {
+  SmallVector<CallInst *, 8> ToReplace;
+  for (Function &F : M) {
+    CallingConv::ID CC = F.getCallingConv();
+
+    for (auto &I : instructions(F)) {
+      auto *CI = dyn_cast<CallInst>(&I);
+      Function *Callee = nullptr;
+      if (!CI || !(Callee = CI->getCalledFunction()))
+        continue;
+      StringRef Name = Callee->getName();
+      if (Name != SYCL_ALLOCLOCALMEM_CALL)
+        continue;
+
+      // TODO: Static local memory allocation should be requested only in
+      // spir kernel scope.
+      assert((CC == llvm::CallingConv::SPIR_FUNC ||
```

Comment:
I see that the logic here is to error out in case of finding SYCL_ALLOCLOCALMEM_CALL in a non-spir function, but still have you considered to move it above to have an early exit in case if the function isn't spir?

Reply:
Then there may be left calls of __sycl_allocateLocalMemory in some non spir function and compilation may fail later with error "unresolved __sycl_allocateLocalMemory", which is not clear for a user.